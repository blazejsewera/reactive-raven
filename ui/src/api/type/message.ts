export type ApiMessage = {
  username: string
  timestamp: string
  title: string
  subtitle?: string
  body?: string
  id?: string
}

// example:
// { "username": "Test", "timestamp": "a", "title": "Test title", "subtitle": "Test subtitle", "body": "Test body.", "id": "1234" }
