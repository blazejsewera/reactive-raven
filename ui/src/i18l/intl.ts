/**
 * Intl is an interface for an internationalization (i18l) tool
 * that can provide translations for page.
 */
export interface Intl {
  getTranslation: (text: string) => string
}

export const intl: Intl = { getTranslation: text => text }
