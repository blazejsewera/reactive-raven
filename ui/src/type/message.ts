export type Message = {
  appName: string
  timestamp: string
  appImgUri?: string
  title: string
  subtitle?: string
  body?: string
  id?: string
}
