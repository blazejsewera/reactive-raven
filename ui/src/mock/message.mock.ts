import { Message } from '../type/message'
import { fiveSentenceLoremIpsum } from './asset/text/lipsum'
import githubIcon from './asset/icon/github-icon.svg'

const body = `This is a simple test message.

It spans over multiple lines.`

export const full: Message = {
  username: 'Primary',
  title: 'Test title',
  subtitle: '— primary',
  body,
  timestamp: '2 hours ago',
  id: '0',
}

export const fullWithImage: Message = {
  ...full,
  id: '1',
  userImgUri: githubIcon,
}

export const fullWithLoremIpsum: Message = {
  ...full,
  id: '2',
  body: fiveSentenceLoremIpsum,
}

export const partial: Message = {
  username: full.username,
  title: full.title,
  body: full.body,
  timestamp: full.timestamp,
  id: '3',
}

export const minimal: Message = {
  username: full.username,
  title: full.title,
  timestamp: full.timestamp,
  id: '4',
}

export const anotherUser: Message = {
  username: 'Other',
  title: 'Other title',
  subtitle: '— other',
  body,
  timestamp: '1 hour ago',
  id: '5',
}
