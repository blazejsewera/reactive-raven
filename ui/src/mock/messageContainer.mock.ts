import type { MessageContainerHandlers } from '../type/handler'

export const handlers: MessageContainerHandlers = {
  onClear: () => {},
}
