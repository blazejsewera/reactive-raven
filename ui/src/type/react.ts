import type { ReactElement, ReactNode } from "react";

type EmptyObject = Record<string, never>;

/**
 * Default children type in React. Every renderable type is included.
 */
export type DefaultChildren = ReactNode;

/**
 * No props, a.k.a. () => </> component
 */
export type EmptyProps = EmptyObject;

/**
 * No React Context
 */
export type EmptyContext = EmptyObject;

/**
 * Functional Component (strongly typed)
 */
export interface FC<P = EmptyProps, Ctx = EmptyContext> {
  (props: P, context?: Ctx): ReactElement | null;
  displayName?: string;
}
