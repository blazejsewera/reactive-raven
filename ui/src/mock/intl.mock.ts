import { Intl } from '../i18l/intl'

export const intlMock: Intl = { getTranslation: text => text }
