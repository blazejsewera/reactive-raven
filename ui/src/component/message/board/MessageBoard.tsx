import * as React from 'react'
import type { Intl } from '../../../i18l/intl'
import type { Message } from '../../../type/message'
import type { FC } from '../../../type/react'
import { cx } from '../../../util/cx'
import { byUsername } from '../../../util/message/categorizer'
import { MessageContainer } from '../container/MessageContainer'

export interface MessageBoardProps {
  messages: Message[]
  intl: Intl
}

export const MessageBoard: FC<MessageBoardProps> = ({ messages, intl }) => {
  const categorized = byUsername(messages)
  const usernames = Object.keys(categorized)
  const containers = usernames.map(username => (
    <MessageContainer
      title={username}
      messages={categorized[username]}
      intl={intl}
      handlers={{ onClear: () => {} }}
      style={cx('inline-block')}
    />
  ))

  return <div className={cx('inline-block space-x-4 whitespace-nowrap')}>{containers}</div>
}
