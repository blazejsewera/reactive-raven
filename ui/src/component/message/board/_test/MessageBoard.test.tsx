import * as React from 'react'
import { create as render } from 'react-test-renderer'
import { intlMock } from '../../../../mock/intl.mock'
import { anotherUser, full, fullWithLoremIpsum } from '../../../../mock/message.mock'
import { handlers as mockContainerHandlers } from '../../../../mock/messageContainer.mock'
import { MessageBoard } from '../MessageBoard'

describe('MessageBoard component', () => {
  it('renders correctly with messages from 2 users', () => {
    // given
    const intl = intlMock
    const messages = [full, fullWithLoremIpsum, anotherUser]
    const containerHandlers = mockContainerHandlers

    // when
    const tree = render(<MessageBoard {...{ messages, intl, containerHandlers }} />).toJSON()

    // then
    expect(tree).toMatchSnapshot()
  })
})
