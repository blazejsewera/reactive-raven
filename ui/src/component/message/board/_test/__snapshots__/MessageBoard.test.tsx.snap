// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBoard component renders correctly with messages from 2 users 1`] = `
<div
  className="inline-block space-x-4 whitespace-nowrap"
>
  <div
    className="bg-gray-100 dark:bg-gray-700 inline-block p-5 pb-6 rounded-3xl shadow-xl inline-block"
  >
    <div
      className="flex mb-4 mx-2"
    >
      <p
        className="text-2xl font-bold text-gray-800 dark:text-gray-300"
      >
        Primary
      </p>
    </div>
    <div
      className="grid grid-cols-1 space-y-5"
    >
      <div
        className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
      >
        <div
          className="flex"
        >
          <div
            className=" h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
          >
            <svg
              viewBox="0 0 1 1"
            >
              <rect
                fill="#111827"
                height={1}
                width={1}
              />
              <text
                dy="0.178"
                fill="#FFF"
                style={
                  Object {
                    "fontFamily": "Inter",
                    "fontSize": 0.5,
                    "fontWeight": "bold",
                  }
                }
                textAnchor="middle"
                x={0.5}
                y={0.5}
              >
                Pr
              </text>
            </svg>
          </div>
          <div
            className="my-auto inline-block ml-5"
          >
            <p
              className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
            >
              Test title
            </p>
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              — primary
            </p>
          </div>
        </div>
        <p
          className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
        >
          <span>
            This is a simple test message.

It spans over multiple lines.
          </span>
        </p>
        <div
          className="flex"
        >
          <div
            className="inline-block"
          >
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              — by 
              Primary
            </p>
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              2 hours ago
            </p>
          </div>
        </div>
      </div>
      <div
        className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
      >
        <div
          className="flex"
        >
          <div
            className=" h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
          >
            <svg
              viewBox="0 0 1 1"
            >
              <rect
                fill="#111827"
                height={1}
                width={1}
              />
              <text
                dy="0.178"
                fill="#FFF"
                style={
                  Object {
                    "fontFamily": "Inter",
                    "fontSize": 0.5,
                    "fontWeight": "bold",
                  }
                }
                textAnchor="middle"
                x={0.5}
                y={0.5}
              >
                Pr
              </text>
            </svg>
          </div>
          <div
            className="my-auto inline-block ml-5"
          >
            <p
              className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
            >
              Test title
            </p>
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              — primary
            </p>
          </div>
        </div>
        <p
          className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
        >
          <span>
            Explicabo ipsum iusto qui porro iste possimus ad voluptatem. Nam maxime voluptatem nostrum deleniti saepe. Officia dolor velit sunt velit placeat amet
            <span
              aria-label="show more"
              className="text-blue-500 dark:text-blue-400 cursor-pointer"
              onClick={[Function]}
            >
              [...]
            </span>
          </span>
        </p>
        <div
          className="flex"
        >
          <div
            className="inline-block"
          >
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              — by 
              Primary
            </p>
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              2 hours ago
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="bg-gray-100 dark:bg-gray-700 inline-block p-5 pb-6 rounded-3xl shadow-xl inline-block"
  >
    <div
      className="flex mb-4 mx-2"
    >
      <p
        className="text-2xl font-bold text-gray-800 dark:text-gray-300"
      >
        Other
      </p>
    </div>
    <div
      className="grid grid-cols-1 space-y-5"
    >
      <div
        className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
      >
        <div
          className="flex"
        >
          <div
            className=" h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
          >
            <svg
              viewBox="0 0 1 1"
            >
              <rect
                fill="#111827"
                height={1}
                width={1}
              />
              <text
                dy="0.178"
                fill="#FFF"
                style={
                  Object {
                    "fontFamily": "Inter",
                    "fontSize": 0.5,
                    "fontWeight": "bold",
                  }
                }
                textAnchor="middle"
                x={0.5}
                y={0.5}
              >
                Ot
              </text>
            </svg>
          </div>
          <div
            className="my-auto inline-block ml-5"
          >
            <p
              className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
            >
              Other title
            </p>
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              — other
            </p>
          </div>
        </div>
        <p
          className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
        >
          <span>
            This is a simple test message.

It spans over multiple lines.
          </span>
        </p>
        <div
          className="flex"
        >
          <div
            className="inline-block"
          >
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              — by 
              Other
            </p>
            <p
              className="text-xs font-bold text-gray-500 dark:text-gray-400"
            >
              1 hour ago
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
